core/src/main/java/amino/run/kernel/client/KernelClient.java

```diff
@@ -28,41 +34,180 @@
+    private static final byte[] randomBytes = new byte[STEP_SIZE_IN_BYTES];
+
+    static {
+        new Random().nextBytes(randomBytes);
+    }
+
+    /* Data holder class used in heartbeats */
+    public static class RandomData implements Serializable {
+        private transient int len = STEP_SIZE_IN_BYTES; /* length of data to be sent to server */
+
+        private void writeObject(java.io.ObjectOutputStream s) throws java.io.IOException {
+            s.writeInt(len);
+            for (int i = 0; i < len; i += STEP_SIZE_IN_BYTES) {
+                /* Write the same byte array again as this data is useless on receiver end */
+                s.write(randomBytes);
+            }
+        }
+
+        private void readObject(java.io.ObjectInputStream s)
+                throws java.io.IOException, ClassNotFoundException {
+            len = s.readInt();
+            for (int i = 0; i < len; i += STEP_SIZE_IN_BYTES) {
+                /* Received data is useless. Just overwrites to same byte array */
+                s.read(randomBytes);
+            }
+        }
+    }
+
     /** Stub for the OMS */
     private OMSServer oms;
-    /** List of hostnames matched to kernel server stubs */
-    private Hashtable<InetSocketAddress, KernelServer> servers;
+
+    /* Maximum number of continuous heartbeat miss allowed */
+    private static final int MAX_HEARTBEAT_MISS_TIMES = 3;
+
+    /* Class to hold remote kernel server info. This class is accessible only within this outer class(KernelClient) */
+    private final class KernelServerInfo {
+        private InetSocketAddress serverAddress; /* Host address of kernel server */
+        private KernelServer remoteRef; /* Remote reference to the kernel server */
+        /* Below reference counts are used to remove a server when it fails to send heartbeats for
+        MAX_HEARTBEAT_MISS_ALLOWED times continuously */
+        /* Global reference count is incremented whenever heartbeat is attempted */
+        private int refCount;
+        private int refCountOnHeartBeat; /* Reference count upon successful heartbeats */
+
+        /* Minimum samples to consider data rates as consistent. */
+        private static final int MIN_STABLE_DATA_RATE_TIMES = 10;
+        private int stableDataRateTimes; /* Number of continuous stable data transfer rates */
+        private RandomData data; /* Amount of random data sent to server in heartbeats */
+
+        private int metricPollPeriod = MIN_METRIC_POLL_PERIOD_MS; /* Poll period */
+        private NodeMetric metric; /* Node metric to server */
+        private ResettableTimer metricsTimer; /* Metric timer for the server */
+
+        private KernelServerInfo(
+                InetSocketAddress serverAddress, KernelServer remoteRef, final NodeMetric metric) {
+            this.serverAddress = serverAddress;
+            this.remoteRef = remoteRef;
+            this.metric = metric;
+            data = new RandomData();
```

Comment:
I don't think that you want to store a whole (possibly very large) buffer of data in memory for every server, even when you're not sending it.  Think of thousands of servers, where for each one you might need to send a few megabytes of data to accurately measure the data rate.   All you need to store is a count of the number of bytes that need to be sent per server.  Then you can generate that data on the fly each time (e.g. by taking an appropriately-sized substring of a common data buffer).

Reply:
Sorry, the above comment is rubbish.  You're not storing the buffer.  Deleting that comment.